### PR TITLE
Make Reexport composable with other import macros

### PR DESCRIPTION
**Pull request by lassepe**
This pull-request is an attempt to make Reexport composable with other import macros. This idea originated from https://github.com/Roger-luo/FromFile.jl/issues/25

In order to facilitate compatibility with FromFile.jl (and potentially other import macro tools) this PR introduces the following change:

1. Any macros in the incoming expression are expanded via `Base.macroexpand`, thus `@reexport` always sees the code as if all the syntax transformations have been written out
2. `@reexport` recursively expands all `:block` expressions  (since macros like `FromFile.@from` return a block of multiple import statements)
3. Reexport is expanded to also handle statements of the form `@reexport import A.b` and `@reexport import A: b` which are transformed to `import A.b; export b` and `import A: b; export b` respectively.

Please let me know what you think about this proposal. I will add tests, once we've worked out if we want this feature at all. Also, please let me know if you prefer this pull quest to be split up into multiple smaller ones.

**Edit:**

The feature number 2 above has the nice side-effect that you can now wrap an entire block in a `@reexport` macro:
```julia
@reexport begin
    using A: b
    import C.d
end
````

### CODE REVIEW

File: src/Reexport.jl
Comment (ararslan) [inline src/Reexport.jl:12]: Rather than removing line numbers, you can skip them during processing, which will preserve them in the transformed expression. One way to do this is to define
```julia
reexport(m::Module, linenum::LineNumberNode) = linenum
```
then remove the `remove_linenums!` and let the `map` below handle this case. Alternatively, you can remove `remove_linenums!` and change the `map` call to something like
```julia
map(e -> e isa LineNumberNode ? e : reexport(m, e), ex.args)
```

File: src/Reexport.jl
Comment (ararslan) [inline src/Reexport.jl:18]:
```suggestion
        Meta.isexpr(ex, :toplevel) && all(e->isa(e, Expr) && e.head == :using, ex.args) ||
        error("@reexport: syntax error")
```
Minor style nit. I find it a bit more difficult to follow when subsequent conditions/side effects aren't indented after the initial condition.

File: src/Reexport.jl
Comment (ararslan) [inline src/Reexport.jl:38]: These lines should be de-indented 4 spaces to match the removal of `esc`.

**Review by ararslan — COMMENTED** [review]
Really nice work! Just a couple of comments but overall I like this idea.

File: src/Reexport.jl
Comment (lassepe) [reply src/Reexport.jl]: I just formatted the whole file with JuliaFormatter.jl. That fixed this formatting plus a few other minor things. Let me know if that formatting is too much, then I'll revert it and only do the changes you suggested here.

Comment (ararslan) [reply src/Reexport.jl]: Yeah, it's uhhhh a bit much 😅

JuliaFormatter with `YASStyle()` would be closer to the existing style used here (though not exactly the same).

File: src/Reexport.jl
Comment (lassepe) [reply src/Reexport.jl]: Thank you for the suggestions, this is a good idea. I added the identity dispatch for `LineNumberNode` which seems most julin.

File: src/Reexport.jl
Comment (lassepe) [reply src/Reexport.jl]: I reverted it and manually indented the things as requested (I hope). Let me know if it is what you want now.

**Comment by lassepe** [comment]
Should I also add some examples to the README?

File: src/Reexport.jl
Comment (ararslan) [reply src/Reexport.jl]: Yep, looks great, thanks!

**Comment by ararslan** [comment]
I've leave that up to you whether you want to add some examples. Examples are great but I'm also fine accepting this PR as-is.

**Comment by lassepe** [comment]
Okay cool. Please don't merge this yet. I think I found another edge-case regarding for `import Foo: bar`. Currently, only `import Foo.bar` is supported

File: src/Reexport.jl
Comment (lassepe) [inline src/Reexport.jl:24]: @ararslan I think this branch can never be reached? In fact, I removed it and all tests still pass. Do you know more?

**Comment by lassepe** [comment]
Another thing that I am wondering: What is the desired behavior of `@reexport import Foo`. Currently it only exports `Foo` not the names within. In my opinion this is the correct behavior because `import Foo` also only adds `Foo` to the current namespace. Or should it be disallowed?

**Comment by lassepe** [comment]
I added examples to the README and also handled `@reexport import Foo: bar`. As for `@reexport import Foo` I think the behavior of not exporting any of Foo's names is correct, so I did not change that. Thus, from my side this would be ready for another review.

File: src/Reexport.jl
Comment (ararslan) [reply src/Reexport.jl]: Ah, you're right, good catch. This package is very old—it was written for Julia 0.2, I believe—and the way some of these expressions are structured internally has since changed but not all of the code here has been updated to match. (I thought it had been but clearly not!) AFAICT, this branch would have accepted `:(using A)` and `:(using A.B)` in earlier Julia versions. As you've noticed, it seems now to be dead code.

**Comment by ararslan** [comment]
> In my opinion this is the correct behavior because `import Foo` also only adds `Foo` to the current namespace.

I agree 👍

File: src/Reexport.jl
Comment (ararslan) [inline src/Reexport.jl:17]: Note that `Meta.isexpr` does a check for `ex isa Expr`, but we know by the signature of this method that `ex` is an `Expr`; it's constrained to be so. I think this could instead be `ex.head in [...]` and likewise in a few other places.

**Review by ararslan — APPROVED** [review]
Great work and thanks for your patience with my review 😄

**Comment by lassepe** [comment]
> Great work and thanks for your patience with my review 😄

Thank you for the detailed review! It's my first time meta programming so it's good to get a lot of feedback :)

**Pull request merged**
